File: easy/020. Avik and his Girlfriend's Necklace/cpp/partition.cpp
#include <iostream>
#include "partition.hpp"

PartitionArray::PartitionArray(int size):total_(0)
{
	array_.resize(size);
}

void PartitionArray::Insert(int number) {
	array_[total_++] = number;
}

int PartitionArray::Partition(int &start, int &end, int &pivot) {
	int left = start - 1;
	int right = end;

	while (true) {
		while (left < end && array_[++left] < array_[pivot]) {
		}
		while (right > start && array_[--right] > array_[pivot]) {
		}

		if (left >= right) {
			break;
		} else {
			Swap(left, right);
		}
	}
	Swap(left, pivot);
	return left;
}

void PartitionArray::Show() {
	for (int i = 0; i < total_; i++) {
		std::cout << array_[i] << (i == total_ - 1 ? "" : " ");
	}
	std::cout << std::endl;
}

void PartitionArray::Sort() {
	SortArray(0, total_);
}

void PartitionArray::SortArray(int start, int end) {
	if (end - start >= 0) {
		int partition = Partition(start, end, end);
		SortArray(start, partition-1);
		SortArray(partition+1, end);
	}
}

void PartitionArray::Swap(int &left, int &right) {
	int temp = array_[left];
	array_[left] = array_[right];
	array_[right] = temp;
}

File: easy/020. Avik and his Girlfriend's Necklace/python/partition.py
class PartitionArray():
	def __init__(self, numbers):
		self.array = numbers[:]

	def Partition(self, start, end, pivot):
		while True:
			left = start
			right = end

			for l in range(left, end+1):
				left = l
				if self.array[l] > self.array[pivot]:
					break

			for r in reversed(range(start, right)):
				right = r
				if self.array[r] < self.array[pivot]:
					break

			if left >= right:
				break
			else:
				self.array[left], self.array[right] = self.array[right], self.array[left]

		self.array[left], self.array[pivot] = self.array[pivot], self.array[left]
		return left

	def Show(self):
		for i in range(len(self.array)):
			print(str(self.array[i]) if i == len(self.array)-1 else str(self.array[i]) + " ", end='')
		print()

	def Sort(self):
		self.SortArray(0, len(self.array)-1)

	def SortArray(self, start, end):
		if end - start >= 0:
			partition = self.Partition(start, end, end)
			self.SortArray(start, partition-1)
			self.SortArray(partition+1, end)

	def Swap(self, left, right):
		temp = self.array[left]
		self.array[left] = self.array[right]
		self.array[right] = temp